File: string/leetcode_205.java
package string;

public class leetcode_205 {
    public static void main(String[] args) {
        leetcode_205 lc = new leetcode_205();
        System.out.println(lc.isIsomorphic("paper", "title"));
    }

    public boolean isIsomorphic(String s, String t) {
        int[] m = new int[512];
        for (int i = 0; i < s.length(); i++) {
            if (m[s.charAt(i)] != m[t.charAt(i) + 256]) return false;
            m[s.charAt(i)] = m[t.charAt(i) + 256] = i + 1;
        }
        return true;
    }
}

File: string/LeetCode_205.kt
package string

class LeetCode_205 {

    fun isIsomorphic(s: String, t: String): Boolean {
        val charMap = hashMapOf<Char, Char>()
        val checked = hashSetOf<Char>()

        for (i in s.indices) {
            val sc = s[i]
            val tc = t[i]

            if (charMap.containsKey(sc)) {
                if (charMap[sc]!! != tc) {
                    return false
                }
            } else if (checked.contains(tc)) {
                return false
            }

            charMap[sc] = tc
            checked.add(tc)
        }

        return true
    }
}

fun main() {
    // Input
    val s = "egg"
    val t = "add"
    // Output
    LeetCode_205().isIsomorphic(s, t).run {
        println(this)
        require(this)
    }
}
